### DISCUSSION TITLE
Date header

### DISCUSSION THREAD

**Discussion by jayvdb** [actix/actix-web#3256] [Ideas] [answered]
According to [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#field.date):

> An origin server with a clock (as defined in [Section 5.6.7](https://www.rfc-editor.org/rfc/rfc9110#http.date)) **MUST** generate a Date header field in all [2xx (Successful)](https://www.rfc-editor.org/rfc/rfc9110#status.2xx), [3xx (Redirection)](https://www.rfc-editor.org/rfc/rfc9110#status.3xx), and [4xx (Client Error)](https://www.rfc-editor.org/rfc/rfc9110#status.4xx) responses, and **MAY** generate a Date header field in [1xx (Informational)](https://www.rfc-editor.org/rfc/rfc9110#status.1xx) and [5xx (Server Error)](https://www.rfc-editor.org/rfc/rfc9110#status.5xx) responses.
> 
> An origin server without a clock **MUST NOT** generate a Date header field.

and Section 5.6.7 is

> A "clock" is an implementation capable of providing a reasonable approximation of the current instant in UTC. A clock implementation ought to use NTP ([[RFC5905](https://www.rfc-editor.org/rfc/rfc9110#RFC5905)]), or some similar protocol, to synchronize with UTC.

actix-web cant be 100% sure that the time it sees (e.g. [use of std::time](https://github.com/search?q=repo%3Aactix%2Factix-web%20std%3A%3Atime&type=code) is using NTP or similar, but it is highly likely, especially given that actix cant build without `std`

Given this is a MUST, IMO it would be great for actix-web to provide an easy way to comply with the spec.

Using [`map_response`](https://github.com/robjtede/actix-web-lab/issues/54) the implementation is roughly

```rust
/// Middleware to add Date header to all responses.
/// Usage: `App::new().wrap(map_response(add_date_header))`
async fn add_date_header(
    mut res: actix_web::dev::ServiceResponse<impl actix_web::body::MessageBody>,
) -> actix_web::Result<actix_web::dev::ServiceResponse<impl actix_web::body::MessageBody>> {
    res.headers_mut().insert(
        actix_web::http::header::DATE,
        actix_web::http::header::HttpDate::from(std::time::SystemTime::now()).try_into_value()?,
    );

    Ok(res)
}
```

**Comment by robjtede** [comment, marked as answer]
Great news, Actix Web already does this automatically!

```
 › curl -v http://127.0.0.1:8080
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080
> GET / HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/8.4.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-length: 6
< date: Wed, 24 Jan 2024 03:52:04 GMT         <-------------------
<
* Connection #0 to host 127.0.0.1 left intact
Hello!%
```

```
 › curl -v --http2-prior-knowledge http://127.0.0.1:8080
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080
* [HTTP/2] [1] OPENED stream for http://127.0.0.1:8080/
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: http]
* [HTTP/2] [1] [:authority: 127.0.0.1:8080]
* [HTTP/2] [1] [:path: /]
* [HTTP/2] [1] [user-agent: curl/8.4.0]
* [HTTP/2] [1] [accept: */*]
> GET / HTTP/2
> Host: 127.0.0.1:8080
> User-Agent: curl/8.4.0
> Accept: */*
>
< HTTP/2 200
< content-length: 6
< date: Wed, 24 Jan 2024 03:52:10 GMT          <-------------------
<
* Connection #0 to host 127.0.0.1 left intact
Hello!%
```

**Reply by jayvdb** [reply]
Thanks for the clue bat.  Apologies for it being a silly question.

I was testing using `actix_web::test::init_service` and found that it wasnt adding a `Date` header.  I've moved on to using `actix_test::start` which does provide the `Date` header.

Perhaps https://actix.rs/docs/testing/ should indicate that the `actix_web::test` interface doesnt include some headers.

**Reply by robjtede** [reply]
Yea there's lots of improvements to be made to the testing docs. I'll remember to include this note as an example of server test != service test.
